test/attribute/attribute_read_io_test.cpp

```diff
@@ -0,0 +1,150 @@
+{};
+
+TEST_F(AttributeReadIO, read_string_scalar_20)
+{
+
+  auto file = file::create("AttributeReadIO.h5",
+			   file::AccessFlags::TRUNCATE);
+  auto r = file.root();
+  std::string data = "hello";
+
+  datatype::String fixed_type = datatype::String::fixed(20);
+
+  auto a = r.attributes.create("scalar20", fixed_type, dataspace::Scalar());
+  a.write(data, fixed_type);
+  std::string buffer;
+  a.read(buffer); // causes terminate called after throwing an instance of 'std::bad_alloc'
+  a.read(buffer, a.datatype());
+  std::cout << buffer << std::endl;
+  file.close();
+}
+
+TEST_F(AttributeReadIO, read_string_scalar_var)
+{
+
+  auto file = file::create("AttributeReadIO.h5",
+			   file::AccessFlags::TRUNCATE);
+  auto r = file.root();
+  std::string data = "hello";
+
+  datatype::String var_type = datatype::String::variable();
+
+  auto a2 = r.attributes.create("scalarvar", var_type, dataspace::Scalar());
+  a2.write(data, var_type);
+  std::string buffer2;
+  a2.read(buffer2);
+  a2.read(buffer2, a2.datatype());
+  std::cout << buffer2 << std::endl;
+  file.close();
+}
+
+TEST_F(AttributeReadIO, read_string_simple_var)
+{
+
+  auto file = file::create("AttributeReadIO.h5",
+			   file::AccessFlags::TRUNCATE);
+  auto r = file.root();
+  std::string data = "hello";
+
+  datatype::String var_type = datatype::String::variable();
+
+  auto a3 = r.attributes.create("simplevar", var_type, dataspace::Simple({1}));
+  a3.write(data, var_type);
+  std::string buffer3;
+  a3.read(buffer3);
+  a3.read(buffer3, a3.datatype());
+  std::cout << buffer3 << std::endl;
+  file.close();
+}
+
+TEST_F(AttributeReadIO, read_string_simple_20)
+{
+
+  auto file = file::create("AttributeReadIO.h5",
+			   file::AccessFlags::TRUNCATE);
+  auto r = file.root();
+  std::string data = "hello";
+
+  datatype::String fixed_type = datatype::String::fixed(20);
+
+  auto a4 = r.attributes.create("simple20", fixed_type, dataspace::Simple({1}));
+  a4.write(data, fixed_type);
+  std::string buffer4;
+  a4.read(buffer4); // causes terminate called after throwing an instance of 'std::bad_alloc'
+  a4.read(buffer4, a4.datatype());
+  std::cout << buffer4 << std::endl;
+
+  file.close();
+}
+
+
+TEST_F(AttributeReadIO, read_string_array_simple_20)
+{
+
+  auto file = file::create("AttributeReadIO.h5",
+			   file::AccessFlags::TRUNCATE);
+  auto r = file.root();
+  std::string data = "hello";
+  std::vector<std::string> vec_data = {"hello", "hello1", "hello2", "hello33"};
+  datatype::String fixed_type = datatype::String::fixed(20);
+
+  auto a5 = r.attributes.create("simple20x", fixed_type, hdf5::dataspace::Simple({4}));
+  a5.write(vec_data, fixed_type);
+  std::vector<std::string> buffer5(4);
+  a5.read(buffer5); // causes terminate called after throwing an instance of 'std::bad_alloc'
+  a5.read(buffer5, a5.datatype());
+  std::cout << buffer5[0] << " " << buffer5[1] << " "
+	<< buffer5[2] << " " << buffer5[3] << " " << std::endl;
+
+  file.close();
+}
+
+
```

Comment:
In general, these tests should probably be in the other files, specifically the ones for string IO and array IO.